packages/core/src/sync/index.ts

```diff
@@ -431,483 +308,487 @@ export const createSync = async (args: CreateSyncParameters): Promise<Sync> => {
-        syncProgress.current = event.block;
-        // Note: this checkpoint is <= the previous checkpoint
-        const checkpoint = getOmnichainCheckpoint("current")!;
-
-        // Update "ponder_sync_block" metric
-        args.common.metrics.ponder_sync_block.set(
-          { network: network.name },
-          hexToNumber(syncProgress.current.number),
-        );
-
-        // Remove all reorged data
-
-        perNetworkSync.get(network)!.unfinalizedBlocks =
-          unfinalizedBlocks.filter(
-            ({ block }) =>
-              hexToNumber(block.number) <= hexToNumber(event.block.number),
-          );
-
-        const isReorgedEvent = ({ chainId, block }: RawEvent) =>
-          chainId === network.chainId &&
-          Number(block.number) > hexToNumber(event.block.number);
+      const syncProgress = await getLocalSyncProgress({
+        common: params.common,
+        network,
+        sources,
+        requestQueue,
+        intervalsCache: historicalSync.intervalsCache,
+      });
 
-        pendingEvents = pendingEvents.filter(
-          (e) => isReorgedEvent(e) === false,
-        );
-        executedEvents = executedEvents.filter(
-          (e) => isReorgedEvent(e) === false,
-        );
+      const realtimeSync = createRealtimeSync({
+        common: params.common,
+        sources,
+        requestQueue,
+        network,
+        onEvent: (event) =>
+          perChainOnRealtimeSyncEvent(event)
+            .then((event) =>
+              onRealtimeSyncEvent(event, {
+                network,
+                syncProgress,
+                realtimeSync,
+              }),
+            )
+            .catch((error) => {
+              params.common.logger.error({
+                service: "sync",
+                msg: `Fatal error: Unable to process ${event.type} event`,
+                error,
+              });
+              params.onFatalError(error);
+            }),
+        onFatalError: params.onFatalError,
+      });
 
-        // Move events from executed to pending
+      params.common.metrics.ponder_sync_is_realtime.set(
+        { network: network.name },
+        0,
+      );
+      params.common.metrics.ponder_sync_is_complete.set(
+        { network: network.name },
+        0,
+      );
 
-        const events = executedEvents.filter((e) => e.checkpoint > checkpoint);
-        executedEvents = executedEvents.filter(
-          (e) => e.checkpoint < checkpoint,
-        );
-        pendingEvents.push(...events);
+      perNetworkSync.set(network, {
+        syncProgress,
+        historicalSync,
+        realtimeSync,
+      });
 
-        await args.syncStore.pruneRpcRequestResult({
-          chainId: network.chainId,
-          blocks: event.reorgedBlocks,
-        });
+      const perChainOnRealtimeSyncEvent = getPerChainOnRealtimeSyncEvent({
+        common: params.common,
+        network,
+        sources,
+        syncStore: params.syncStore,
+        syncProgress,
+        realtimeSync,
```

Comment:
It's a bummer that the `perChainOnRealtimeSyncEvent` needs the `realtimeSync`, creates a bit of indirection.